waspc/packages/deploy/src/common/waspBuild.ts

```diff
@@ -0,0 +1,31 @@
+import { WaspCliExe, WaspProjectDir } from "./brandedTypes.js";
+import { waspSays } from "./terminal.js";
+import { createCommandWithCwd } from "./zx.js";
+
+export const ensureWaspProjectIsBuilt = createEnsureWaspProjectIsBuilt();
+
+function createEnsureWaspProjectIsBuilt() {
+  // We want to build the Wasp project only once per CLI invocation.
+  // Sometimes `ensureWaspProjectIsBuilt` is called multiple times
+  // (e.g. in `setup` command, we call it and then again when deploying).
+  let isWaspBuildExecuted = false;
+
+  async function ensureWaspProjectIsBuilt({
+    waspProjectDir,
+    waspExe,
+  }: {
+    waspProjectDir: WaspProjectDir;
+    waspExe: WaspCliExe;
+  }): Promise<void> {
+    if (isWaspBuildExecuted) {
+      return;
+    }
```

Comment:
This has a race condition:

```js
const taskA = async () => {
  await ensureWaspProjectIsBuilt()
  await dependsOnBuildA()
}

const taskB = async () => {
  await ensureWaspProjectIsBuilt()
  await dependsOnBuildB()
}

await Promise.all([taskA(), taskB()])
```

Both `taskA` and `taskB` will interfere build and interfere between each other, with one of the tasks possibly going ahead while the other task is in the midst of a re-build again.

Proposed solution: instead of storing a boolean, immediately store the actual result promise and always return that.